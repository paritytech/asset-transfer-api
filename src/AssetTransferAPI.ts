import { ApiPromise } from '@polkadot/api';
import { Bytes, Option, u32 } from '@polkadot/types';
import { MultiLocation } from '@polkadot/types/interfaces';

interface ITransferArgsOpts {
	/**
	 * Signing Payload vs Call
	 */
	format?: 'payload' | 'call'; // Give a polkadot-js option.
	/**
	 * AssetId to pay fee's on the current common good parachain.
	 * Statemint: default DOT
	 * Statemine: default KSM
	 */
	payFeeWith?: string;
	/**
	 * AssetId to pay fee's on the destination parachain.
	 */
	payFeeWithTo?: string;
	/**
	 * Boolean to declare if this will be with limited XCM transfers.
	 * Deafult is unlimited.
	 */
	isLimited?: boolean;
}

interface IChainInfo {
	specName: string;
	specVersion: string;
}

export class AssetsTransferAPI {
	readonly _api: ApiPromise;
	readonly _info: Promise<IChainInfo>;

	constructor(api: ApiPromise) {
		this._api = api;
		this._info = this.fetchChainInfo(api);
	}

	/**
	 * TBD
	 *
	 * @param destChainId ID of the destination (para) chain (‘0’ for Relaychain)
	 * @param destAddr Address of destination account
	 * @param assetId ID of the asset to be transferred (‘0’ for DOT)
	 * @param amount Amount of the token to transfer
	 * @param opts Options
	 */
	public async createTransferTransaction(
		destChainId: string | number,
		destAddr: string,
		assetId: string,
		amount: string | number,
		opts?: ITransferArgsOpts
	) {
		const { _api, _info } = this;
		const { specName, specVersion } = await _info;
		console.log(
			destChainId,
			destAddr,
			assetId,
			amount,
			opts,
			specName,
			specVersion
		);
		/**
		 * `api.tx.xcmPallets` methods to support inlcude:
		 *	 'teleportAssets',
		 *	 'reserveTransferAssets',
		 *	 'limitedReserveTransferAssets',
		 *	 'limitedTeleportAssets'
		 */
		console.log(Object.keys(_api.tx.xcmPallet));
	}

	/**
	 * Return a partialFee of the
	 *
	 * @param tx Transaction to estimate the fee for
	 */
	public estimateFee(tx: Bytes | string) {
		console.log(tx);
	}

	/**
	 * Fetch runtime information based on the connected chain.
	 *
	 * @param api ApiPromise
	 */
	private async fetchChainInfo(api: ApiPromise): Promise<IChainInfo> {
		const { specName, specVersion } = await api.rpc.state.getRuntimeVersion();
		return {
			specName: specName.toString(),
			specVersion: specVersion.toString(),
		};
	}

	/**
	 * TODO: When we are actively using this change it over to `private`.
	 * Fetch the xcmVersion to use for a given chain. If the supported version doesn't for
	 * a given destination we use the on storage safe version.
	 *
	 * @param xcmVersion The version we want to see is supported
	 * @param multiLocation Destination multilocation
	 */
	public async fetchXcmVersion(
		xcmVersion: number,
		multiLocation: MultiLocation,
		fallbackVersion: number
	): Promise<number | u32> {
		const { _api } = this;

		const supportedVersion = await _api.query.polkadotXcm.supportedVersion<
			Option<u32>
		>(xcmVersion, multiLocation);

		if (supportedVersion.isNone) {
			const safeVersion = await _api.query.polkadotXcm.safeXcmVersion<
				Option<u32>
			>();
			const version = safeVersion.isSome
				? safeVersion.unwrap()
				: fallbackVersion;
			return version;
		}

		return supportedVersion.unwrap();
	}
}
