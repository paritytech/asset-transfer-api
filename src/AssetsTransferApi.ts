// Copyright 2023 Parity Technologies (UK) Ltd.

import '@polkadot/api-augment';

import type { ApiPromise } from '@polkadot/api';
import type { SubmittableExtrinsic } from '@polkadot/api/submittable/types';
import type { Option, u32 } from '@polkadot/types';
import type { ISubmittableResult } from '@polkadot/types/types';

import {
	DEFAULT_XCM_VERSION,
	SYSTEM_PARACHAINS_IDS,
	SYSTEM_PARACHAINS_NAMES,
} from './consts';
import { transfer, transferKeepAlive } from './createCalls';
import {
	limitedReserveTransferAssets,
	reserveTransferAssets,
} from './createXcmCalls';
import { establishXcmPallet } from './createXcmCalls/util/establishXcmPallet';
import { checkLocalTxInput, checkXcmVersion } from './errors';
import { parseRegistry } from './registry/parseRegistry';
import type { ChainInfoRegistry } from './registry/types';
import { sanitizeAddress } from './sanitize/sanitizeAddress';
import {
	ConstructedFormat,
	Format,
	IAssetsTransferApiOpts,
	IChainInfo,
	IDirection,
	IMethods,
	ITransferArgsOpts,
	TxResult,
} from './types';

export class AssetsTransferApi {
	readonly _api: ApiPromise;
	readonly _opts: IAssetsTransferApiOpts;
	readonly _info: Promise<IChainInfo>;
	readonly _safeXcmVersion: Promise<u32>;
	readonly _registry: ChainInfoRegistry;

	constructor(api: ApiPromise, opts: IAssetsTransferApiOpts = {}) {
		this._api = api;
		this._opts = opts;
		this._info = this.fetchChainInfo();
		this._safeXcmVersion = this.fetchSafeXcmVersion();
		this._registry = parseRegistry(opts);
	}

	/**
	 * Create an asset transfer transaction. This can be either locally on a systems parachain,
	 * or between chains using xcm.
	 *
	 * @param destChainId ID of the destination (para) chain (‘0’ for Relaychain)
	 * @param destAddr Address of destination account
	 * @param assetIds Array of assetId's to be transferred (‘0’ for Native Relay Token)
	 * @param amounts Array of the amounts of each token to transfer
	 * @param opts Options
	 */
	public async createTransferTransaction<T extends Format>(
		destChainId: string,
		destAddr: string,
		assetIds: string[],
		amounts: string[],
		opts?: ITransferArgsOpts<T>
	): Promise<TxResult<T>> {
		const { _api, _info, _safeXcmVersion } = this;
		const { specName } = await _info;
		const safeXcmVersion = await _safeXcmVersion;
		const isOriginSystemParachain = SYSTEM_PARACHAINS_NAMES.includes(
			specName.toLowerCase()
		);
		const isDestSystemParachain = SYSTEM_PARACHAINS_IDS.includes(destChainId);

		/**
		 * Sanitize the address to a hex, and ensure that the passed in SS58, or publickey
		 * is validated correctly.
		 */
		const addr = sanitizeAddress(destAddr);
		/**
		 * Create a local asset transfer.
		 */
		if (isDestSystemParachain && isOriginSystemParachain) {
			/**
			 * This will throw a BaseError if the inputs are incorrect and don't
			 * fit the constraints for creating a local asset transfer.
			 */
			checkLocalTxInput(assetIds, amounts); // Throws an error when any of the inputs are incorrect.
			const method = opts?.keepAlive ? 'transferKeepAlive' : 'transfer';
			const tx =
				method === 'transferKeepAlive'
					? transferKeepAlive(_api, addr, assetIds[0], amounts[0])
					: transfer(_api, addr, assetIds[0], amounts[0]);

			return this.constructFormat(tx, 'local', null, method, opts?.format);
		}

		/**
		 * Establish the Transaction Direction
		 */
		const xcmDirection = this.establishDirection(destChainId, specName);
		const isRelayDirection = xcmDirection.toLowerCase().includes('relay');
		/**
		 * Establish an xcmVersion
		 */
		const xcmVersion =
			opts?.xcmVersion === undefined
				? safeXcmVersion.toNumber()
				: opts.xcmVersion;
		checkXcmVersion(xcmVersion); // Throws an error when the xcmVersion is not supported.

		// TODO: Check for xcm construction errors depending on the input.

		/**
		 * Lengths should match, and indicies between both the amounts and assetIds should match.
		 */
		if (assetIds.length !== amounts.length && !isRelayDirection) {
			throw Error(
				'`amounts`, and `assetIds` fields should match in length when constructing a tx from a parachain to a parachain or locally on a system parachain.'
			);
		}

		let txMethod: IMethods;
		let transaction: SubmittableExtrinsic<'promise', ISubmittableResult>;
		if (opts?.isLimited) {
			txMethod = 'limitedReserveTransferAssets';
			transaction = limitedReserveTransferAssets(
				_api,
				xcmDirection,
				addr,
				assetIds,
				amounts,
				destChainId,
				xcmVersion,
				opts?.weightLimit
			);
		} else {
			txMethod = 'reserveTransferAssets';
			transaction = reserveTransferAssets(
				_api,
				xcmDirection,
				addr,
				assetIds,
				amounts,
				destChainId,
				xcmVersion
			);
		}

		return this.constructFormat<T>(
			transaction,
			xcmDirection,
			xcmVersion,
			txMethod,
			opts?.format
		);
	}

	/**
	 * Fetch runtime information based on the connected chain.
	 *
	 * @param api ApiPromise
	 */
	private async fetchChainInfo(): Promise<IChainInfo> {
		const { _api } = this;
		const { specName, specVersion } = await _api.rpc.state.getRuntimeVersion();
		return {
			specName: specName.toString(),
			specVersion: specVersion.toString(),
		};
	}

	/**
	 * Declare the direction of the xcm message.
	 *
	 * @param destChainId
	 * @param specName
	 */
	private establishDirection(
		destChainId: string,
		specName: string
	): IDirection {
		const { _api } = this;
		const isSystemParachain = SYSTEM_PARACHAINS_NAMES.includes(
			specName.toLowerCase()
		);
		const isDestIdSystemPara = SYSTEM_PARACHAINS_IDS.includes(destChainId);

		/**
		 * Check if the origin is a System Parachain
		 */
		if (isSystemParachain && destChainId === '0') {
			return IDirection.SystemToRelay;
		}

		if (isSystemParachain && destChainId !== '0') {
			return IDirection.SystemToPara;
		}

		/**
		 * Check if the origin is a Relay Chain
		 */
		if (_api.query.paras && isDestIdSystemPara) {
			return IDirection.RelayToSystem;
		}

		if (_api.query.paras && !isDestIdSystemPara) {
			return IDirection.RelayToPara;
		}

		/**
		 * Check if the origin is a Parachain or Parathread
		 */
		if (_api.query.polkadotXcm && !isDestIdSystemPara) {
			throw Error('ParaToRelay is not yet implemented');

			return IDirection.ParaToRelay;
		}

		if (_api.query.polkadotXcm) {
			throw Error('ParaToPara is not yet implemented');

			return IDirection.ParaToPara;
		}

		throw Error('Could not establish a xcm transaction direction');
	}

	/**
	 * Construct the correct format for the transaction.
	 * If nothing is passed in, the format will default to a signing payload.
	 *
	 * @param tx A polkadot-js submittable extrinsic
	 * @param format The format to return the tx in.
	 */
	private constructFormat<T extends Format>(
		tx: SubmittableExtrinsic<'promise', ISubmittableResult>,
		direction: IDirection | 'local',
		xcmVersion: number | null = null,
		method: IMethods,
		format?: T
	): TxResult<T> {
		const { _api } = this;
		const fmt = format ? format : 'payload';
		const result: TxResult<T> = {
			direction,
			xcmVersion,
			method,
			format: fmt as Format | 'local',
			tx: '' as ConstructedFormat<T>,
		};

		if (fmt === 'call') {
			result.tx = _api.registry
				.createType('Call', {
					callIndex: tx.callIndex,
					args: tx.args,
				})
				.toHex() as ConstructedFormat<T>;
		}

		if (fmt === 'submittable') {
			result.tx = tx as ConstructedFormat<T>;
		}

		if (fmt === 'payload') {
			result.tx = _api.registry
				.createType('ExtrinsicPayload', tx, {
					version: tx.version,
				})
				.toHex() as ConstructedFormat<T>;
		}

		return result;
	}

	/**
	 * Fetch for a safe Xcm Version from the chain, if none exists the
	 * in app default version will be used.
	 */
	private async fetchSafeXcmVersion(): Promise<u32> {
		const { _api } = this;
		const pallet = establishXcmPallet(_api);
		const safeVersion = await _api.query[pallet].safeXcmVersion<Option<u32>>();
		const version = safeVersion.isSome
			? safeVersion.unwrap()
			: _api.registry.createType('u32', DEFAULT_XCM_VERSION);

		return version;
	}
}
