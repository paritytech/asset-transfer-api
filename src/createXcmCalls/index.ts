// Copyright 2024 Parity Technologies (UK) Ltd.

export { claimAssets } from './polkadotXcm/claimAssets';
export { limitedReserveTransferAssets } from './polkadotXcm/limitedReserveTransferAssets';
export { limitedTeleportAssets } from './polkadotXcm/limitedTeleportAssets';
export { transferAssets } from './polkadotXcm/transferAssets';
export { transferAssetsUsingTypeAndThen } from './polkadotXcm/transferAssetsUsingTypeAndThen';
export { transferMultiasset } from './xTokens/transferMultiasset';
export { transferMultiassets } from './xTokens/transferMultiassets';
export { transferMultiassetWithFee } from './xTokens/transferMultiassetWithFee';
