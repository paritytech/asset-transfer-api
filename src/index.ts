// Copyright 2024 Parity Technologies (UK) Ltd.

export * from './AssetTransferApi';
export * from './constructApiPromise';
export * from './types';
