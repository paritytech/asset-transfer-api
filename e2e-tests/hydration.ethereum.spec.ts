import { setupNetworks, testingPairs, withExpect } from '@acala-network/chopsticks-testing';
import { NetworkContext } from '@acala-network/chopsticks-utils';
import { setTimeout } from 'timers/promises';
import { afterEach, beforeEach, expect, test } from 'vitest';

import { AssetTransferApi } from '../src/AssetTransferApi';

const { checkSystemEvents } = withExpect(expect);

describe('Hydration <> Ethereum', () => {
	const ethereumNetworkGlobalConsensusLocation = `{"parents":"2","interior":{"X1":{"GlobalConsensus":{"Ethereum":{"chainId":"1"}}}}}`;
	let hydration: NetworkContext;
	let polkadotAssetHub: NetworkContext;
	let polkadotBridgeHub: NetworkContext;

	const { alice, alith } = testingPairs();

	beforeEach(async () => {
		const { hydration1, polkadotBridgeHub1, polkadotAssetHub1 } = await setupNetworks({
			hydration1: {
				endpoint: 'wss://rpc.hydradx.cloud',
				db: './db.sqlite',
				port: 8010,
			},
			polkadotBridgeHub1: {
				endpoint: 'wss://bridge-hub-polkadot-rpc.dwellir.com',
				db: './db.sqlite',
				port: 8011,
			},
			polkadotAssetHub1: {
				endpoint: 'wss://asset-hub-polkadot-rpc.dwellir.com',
				db: './db.sqlite',
				port: 8012,
			},
		});

		hydration = hydration1;
		polkadotBridgeHub = polkadotBridgeHub1;
		polkadotAssetHub = polkadotAssetHub1;
	}, 1000000);

	afterEach(async () => {
		await hydration.teardown();
		await polkadotAssetHub.teardown();
		await polkadotBridgeHub.teardown();
	}, 1000000);

	describe('XCM V3', () => {
		const xcmVersion = 3;

		test('Transfer Snowbridge WETH from Hydration to Ethereum', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
						[[alice.address, 5], { free: '50000000000000000000000000' }], // DOT
						[[alice.address, 1000189], { free: '500000000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2034: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'hydradx',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: '1000189',
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});

			const tx = await assetTransferApi.createTransferTransaction(
				ethereumNetworkGlobalConsensusLocation,
				alith.address,
				['DOT', 'WETH.snow'],
				['500000000000', '75000000000000'],
				{
					sendersAddr: alice.address,
					format: 'payload',
					xcmVersion,
					paysWithFeeDest: 'DOT',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();

			await checkSystemEvents(hydration, 'polkadotXcm')
				.redact({ redactKeys: new RegExp('messageId') })
				.toMatchSnapshot('hydration xcm message sent');

			await setTimeout(10000);
			await polkadotAssetHub.dev.timeTravel(1);

			await checkSystemEvents(polkadotAssetHub, 'foreignAssets').toMatchSnapshot('assethub foreign assets burned');
			await checkSystemEvents(polkadotAssetHub, 'xcmpQueue', 'XcmpMessageSent').toMatchSnapshot(
				'assetHub xcmp message sent',
			);

			await setTimeout(10000);
			await polkadotBridgeHub.dev.timeTravel(1);

			await checkSystemEvents(polkadotBridgeHub, 'ethereumOutboundQueue')
				.redact({ redactKeys: new RegExp('nonce') })
				.toMatchSnapshot('bridgehub ethereum outbound queue events');
		}, 200000);
	});
	describe('XCM V4', () => {
		const xcmVersion = 4;

		test('Transfer Snowbridge WETH from Hydration to Ethereum', async () => {
			await hydration.dev.setStorage({
				System: {
					Account: [
						[[alice.address], { providers: 1, data: { free: 10 * 1e12 } }], // HDX
					],
				},
				Tokens: {
					Accounts: [
						[[alice.address, 0], { free: '50000000000000000000000000' }], // HDX
						[[alice.address, 5], { free: '50000000000000000000000000' }], // DOT
						[[alice.address, 1000189], { free: '500000000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const assetTransferApi = new AssetTransferApi(hydration.api, 'hydradx', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2034: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'hydradx',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: '1000189',
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});

			const tx = await assetTransferApi.createTransferTransaction(
				ethereumNetworkGlobalConsensusLocation,
				alith.address,
				['DOT', 'WETH.snow'],
				['500000000000', '75000000000000'],
				{
					sendersAddr: alice.address,
					format: 'payload',
					xcmVersion,
					paysWithFeeDest: 'DOT',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await hydration.api.tx(extrinsic).signAndSend(alice);
			await hydration.dev.newBlock();

			await checkSystemEvents(hydration, 'polkadotXcm')
				.redact({ redactKeys: new RegExp('messageId') })
				.toMatchSnapshot('hydration xcm message sent');

			await setTimeout(10000);
			await polkadotAssetHub.dev.timeTravel(1);

			await checkSystemEvents(polkadotAssetHub, 'foreignAssets').toMatchSnapshot('assethub foreign assets burned');
			await checkSystemEvents(polkadotAssetHub, 'xcmpQueue', 'XcmpMessageSent').toMatchSnapshot(
				'assetHub xcmp message sent',
			);

			const assetHubEvents = await polkadotAssetHub.api.query.system.events();

			const xcmMessageProcessed = assetHubEvents[assetHubEvents.length - 1];
			expect(xcmMessageProcessed.phase.toString()).toEqual('Finalization');
			expect(xcmMessageProcessed.event.method).toEqual('Processed');
			expect(xcmMessageProcessed.event.section).toEqual('messageQueue');

			await setTimeout(10000);
			await polkadotBridgeHub.dev.timeTravel(1);

			await checkSystemEvents(polkadotBridgeHub, 'ethereumOutboundQueue')
				.redact({ redactKeys: new RegExp('nonce') })
				.toMatchSnapshot('bridgehub ethereum outbound queue events');

			const bridgeHubEvents = await polkadotBridgeHub.api.query.system.events();
			const messageAcceptedEvent = bridgeHubEvents[bridgeHubEvents.length - 3];
			expect(messageAcceptedEvent.event.section).toEqual('ethereumOutboundQueue');
			expect(messageAcceptedEvent.event.method).toEqual('MessageAccepted');

			const messageCommittedEvent = bridgeHubEvents[bridgeHubEvents.length - 1];
			expect(messageCommittedEvent.event.section).toEqual('ethereumOutboundQueue');
			expect(messageCommittedEvent.event.method).toEqual('MessagesCommitted');
		}, 200000);
	});
});
