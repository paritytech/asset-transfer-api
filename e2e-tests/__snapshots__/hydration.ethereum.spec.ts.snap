// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hydration <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Hydration to Ethereum > assetHub xcmp message sent 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Hydration to Ethereum > assethub foreign assets burned 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": {
          "X2": [
            {
              "GlobalConsensus": {
                "Ethereum": {
                  "chainId": 1,
                },
              },
            },
            {
              "AccountKey20": {
                "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "network": null,
              },
            },
          ],
        },
        "parents": 2,
      },
      "balance": 75000000000000,
      "owner": "13cKp89Uh2yWgTG28JA1QEvPUMjEPKejqkjHKf9zqLiFKjH6",
    },
    "method": "Burned",
    "section": "foreignAssets",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Hydration to Ethereum > bridgehub ethereum outbound queue events 1`] = `
[
  {
    "data": {
      "id": 0,
    },
    "method": "MessageQueued",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "id": 0,
      "nonce": "(redacted)",
    },
    "method": "MessageAccepted",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "count": 1,
      "root": "(hash)",
    },
    "method": "MessagesCommitted",
    "section": "ethereumOutboundQueue",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Hydration to Ethereum > hydration xcm message sent 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(redacted)",
            "refTime": "(redacted)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "fees": [],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 500000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            {
              "fun": {
                "Fungible": 75000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 500000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "SetAppendix": [
            {
              "DepositAsset": {
                "assets": {
                  "Wild": "All",
                },
                "beneficiary": {
                  "interior": {
                    "X1": [
                      {
                        "AccountId32": {
                          "id": "(hash)",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 0,
                },
              },
            },
          ],
        },
        {
          "InitiateReserveWithdraw": {
            "assets": {
              "Wild": {
                "AllOf": {
                  "fun": "Fungible",
                  "id": {
                    "interior": {
                      "X2": [
                        {
                          "GlobalConsensus": {
                            "Ethereum": {
                              "chainId": 1,
                            },
                          },
                        },
                        {
                          "AccountKey20": {
                            "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 2,
                  },
                },
              },
            },
            "reserve": {
              "interior": {
                "X1": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                ],
              },
              "parents": 2,
            },
            "xcm": [
              {
                "BuyExecution": {
                  "fees": {
                    "fun": {
                      "Fungible": 1,
                    },
                    "id": {
                      "interior": {
                        "X1": [
                          {
                            "AccountKey20": {
                              "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                              "network": null,
                            },
                          },
                        ],
                      },
                      "parents": 0,
                    },
                  },
                  "weightLimit": "Unlimited",
                },
              },
              {
                "DepositAsset": {
                  "assets": {
                    "Wild": {
                      "AllCounted": 1,
                    },
                  },
                  "beneficiary": {
                    "interior": {
                      "X1": [
                        {
                          "AccountKey20": {
                            "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "SetTopic": 0,
              },
            ],
          },
        },
        {
          "SetTopic": 0,
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Hydration to Ethereum > assetHub xcmp message sent 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Hydration to Ethereum > assethub foreign assets burned 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": {
          "X2": [
            {
              "GlobalConsensus": {
                "Ethereum": {
                  "chainId": 1,
                },
              },
            },
            {
              "AccountKey20": {
                "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "network": null,
              },
            },
          ],
        },
        "parents": 2,
      },
      "balance": 75000000000000,
      "owner": "13cKp89Uh2yWgTG28JA1QEvPUMjEPKejqkjHKf9zqLiFKjH6",
    },
    "method": "Burned",
    "section": "foreignAssets",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Hydration to Ethereum > bridgehub ethereum outbound queue events 1`] = `
[
  {
    "data": {
      "id": 0,
    },
    "method": "MessageQueued",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "id": 0,
      "nonce": "(redacted)",
    },
    "method": "MessageAccepted",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "count": 1,
      "root": "(hash)",
    },
    "method": "MessagesCommitted",
    "section": "ethereumOutboundQueue",
  },
]
`;

exports[`Hydration <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Hydration to Ethereum > hydration xcm message sent 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(redacted)",
            "refTime": "(redacted)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "fees": [],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 500000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            {
              "fun": {
                "Fungible": 75000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 500000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "SetAppendix": [
            {
              "DepositAsset": {
                "assets": {
                  "Wild": "All",
                },
                "beneficiary": {
                  "interior": {
                    "X1": [
                      {
                        "AccountId32": {
                          "id": "(hash)",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 0,
                },
              },
            },
          ],
        },
        {
          "InitiateReserveWithdraw": {
            "assets": {
              "Wild": {
                "AllOf": {
                  "fun": "Fungible",
                  "id": {
                    "interior": {
                      "X2": [
                        {
                          "GlobalConsensus": {
                            "Ethereum": {
                              "chainId": 1,
                            },
                          },
                        },
                        {
                          "AccountKey20": {
                            "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 2,
                  },
                },
              },
            },
            "reserve": {
              "interior": {
                "X1": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                ],
              },
              "parents": 2,
            },
            "xcm": [
              {
                "BuyExecution": {
                  "fees": {
                    "fun": {
                      "Fungible": 1,
                    },
                    "id": {
                      "interior": {
                        "X1": [
                          {
                            "AccountKey20": {
                              "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                              "network": null,
                            },
                          },
                        ],
                      },
                      "parents": 0,
                    },
                  },
                  "weightLimit": "Unlimited",
                },
              },
              {
                "DepositAsset": {
                  "assets": {
                    "Wild": {
                      "AllCounted": 1,
                    },
                  },
                  "beneficiary": {
                    "interior": {
                      "X1": [
                        {
                          "AccountKey20": {
                            "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "SetTopic": 0,
              },
            ],
          },
        },
        {
          "SetTopic": 0,
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountId32": {
                "id": "(hash)",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;
